// @neuledge/states/dist/index.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/arguments.ts
import { ParsingError } from "@neuledge/states-parser";
var parseArguments = /* @__PURE__ */ __name((nodes, parser) => {
  const acc = {};
  for (const node of nodes) {
    if (node.key.name in acc) {
      throw new ParsingError(node.key, `Duplicate argument '${node.key.name}'`);
    }
    acc[node.key.name] = parseArgument(node, parser);
  }
  return acc;
}, "parseArguments");
var parseArgument = /* @__PURE__ */ __name((node, parser) => ({
  type: "Argument",
  node,
  name: node.key.name,
  value: parser(node.value)
}), "parseArgument");

// src/context.ts
import { ParsingError as ParsingError14, STATE_FIELD_INDEX_MAX_INPUT_VALUE, parseStates } from "@neuledge/states-parser";

// src/either.ts
import { ParsingError as ParsingError3 } from "@neuledge/states-parser";
import { z } from "zod";

// src/decorators.ts
import { ParsingError as ParsingError2 } from "@neuledge/states-parser";
var createDecorator = /* @__PURE__ */ __name((decoratorArgs, decorator) => (target, node) => {
  const argsNodes = Object.fromEntries(node.arguments.map((arg) => [
    arg.key.name,
    arg
  ]));
  const argsValues = Object.fromEntries(Object.entries(argsNodes).map(([key, value]) => [
    key,
    value.value?.value
  ]));
  const args = decoratorArgs.safeParse(argsValues);
  if (!args.success) {
    const [issue] = args.error.issues;
    const key = issue.path[0];
    throw new ParsingError2(argsNodes[key] ?? node.callee, `Invalid '@${node.callee.name}()' decorator on argument '${key}': ${issue.message}`);
  }
  return decorator(target, args.data, argsNodes, node);
}, "createDecorator");
var applyDecorators = /* @__PURE__ */ __name((target, nodes, decorators6) => {
  for (const node of nodes) {
    const { name } = node.callee;
    const decorator = decorators6[name];
    if (decorator) {
      decorator(target, node);
    }
  }
}, "applyDecorators");

// src/either.ts
var parseEither = /* @__PURE__ */ __name((ctx, node) => {
  const either = {
    type: "Either",
    node,
    name: node.id.name,
    description: node.description?.value,
    states: node.states.map((identifier) => {
      const state = ctx.entity(identifier.name);
      if (state?.type !== "State") {
        throw new ParsingError3(identifier, `Unknown state name '${identifier.name}'`);
      }
      return state;
    })
  };
  applyDecorators(either, node.decorators, decorators);
  return either;
}, "parseEither");
var decorators = {
  deprecated: createDecorator(z.object({
    reason: z.string().optional()
  }), (state, args) => {
    state.deprecated = args.reason || true;
  })
};

// src/mutation/index.ts
import { ParsingError as ParsingError11 } from "@neuledge/states-parser";
import { z as z3 } from "zod";

// src/type.ts
import { ParsingError as ParsingError5 } from "@neuledge/states-parser";

// src/entity.ts
import { ParsingError as ParsingError4 } from "@neuledge/states-parser";
var parseEntity = /* @__PURE__ */ __name((ctx, node) => {
  const entity = ctx.entity(node.name);
  if (!entity) {
    throw new ParsingError4(node, `Unknown entity name '${node.name}'`);
  }
  return entity;
}, "parseEntity");
var parseNonNullableEntity = /* @__PURE__ */ __name((ctx, node) => {
  const entity = ctx.entity(node.name);
  if (!entity) {
    throw new ParsingError4(node, `Unknown entity name '${node.name}'`);
  }
  if (entity.type === "Void") {
    throw new ParsingError4(node, `Void entity is nullable`);
  }
  return entity;
}, "parseNonNullableEntity");

// src/type.ts
var parseType = /* @__PURE__ */ __name((ctx, node) => {
  if (node.type === "TypeGenerator") {
    throw new ParsingError5(node, "Not implemented");
  }
  return parseEntityExpression(ctx, node);
}, "parseType");
var parseEntityExpression = /* @__PURE__ */ __name((ctx, node) => ({
  type: "EntityExpression",
  node,
  entity: parseNonNullableEntity(ctx, node.identifier),
  list: node.list
}), "parseEntityExpression");

// src/parameter.ts
import { ParsingError as ParsingError6 } from "@neuledge/states-parser";
import { z as z2 } from "zod";
var parseParametersContext = /* @__PURE__ */ __name((ctx, target, nodes) => ({
  this: target,
  parameters: parseParameters(ctx, nodes)
}), "parseParametersContext");
var parseParameters = /* @__PURE__ */ __name((ctx, nodes) => {
  const parameters = {};
  for (const node of nodes) {
    if (node.key.name in parameters) {
      throw new ParsingError6(node.key, `Duplicate parameter name '${node.key.name}'`);
    }
    parameters[node.key.name] = parseParameter(ctx, node);
  }
  return parameters;
}, "parseParameters");
var parseParameter = /* @__PURE__ */ __name((ctx, node) => {
  const parameter = {
    type: "Parameter",
    node,
    name: node.key.name,
    description: node.description?.value,
    nullable: node.nullable,
    as: parseType(ctx, node.as)
  };
  applyDecorators(parameter, node.decorators, decorators2);
  return parameter;
}, "parseParameter");
var decorators2 = {
  deprecated: createDecorator(z2.object({
    reason: z2.string().optional()
  }), (parameter, args) => {
    parameter.deprecated = args.reason || true;
  })
};

// src/property.ts
import { ParsingError as ParsingError10 } from "@neuledge/states-parser";

// src/literal.ts
var parseLiteral = /* @__PURE__ */ __name((node) => ({
  type: "Literal",
  node,
  value: node.value
}), "parseLiteral");
var parseNullLiteral = /* @__PURE__ */ __name((node) => ({
  type: "NullLiteral",
  node,
  value: null
}), "parseNullLiteral");

// src/expression/call.ts
import { runtime } from "@neuledge/scalars";
import { ParsingError as ParsingError7 } from "@neuledge/states-parser";
var parseCallExpression = /* @__PURE__ */ __name((params, node) => {
  const callee = runtime[node.callee.name];
  if (!callee) {
    throw new ParsingError7(node.callee, `Unknown runtime function '${node.callee.name}'`);
  }
  return {
    type: "CallExpression",
    node,
    callee,
    arguments: parseArguments(node.arguments, (node2) => parseExpression(params, node2))
  };
}, "parseCallExpression");

// src/expression/identifier.ts
import { ParsingError as ParsingError8 } from "@neuledge/states-parser";
var parseIdentifierExpression = /* @__PURE__ */ __name((params, node) => {
  const reference = params.parameters[node.name];
  if (!reference) {
    throw new ParsingError8(node, `Unknown identifier '${node.name}'`);
  }
  return {
    type: "IdentifierExpression",
    node,
    reference
  };
}, "parseIdentifierExpression");

// src/expression/logical.ts
var parseLogicalExpression = /* @__PURE__ */ __name((params, node) => ({
  type: "LogicalExpression",
  node,
  operator: node.operator,
  left: parseExpression(params, node.left),
  right: parseExpression(params, node.right)
}), "parseLogicalExpression");

// src/expression/this.ts
import { ParsingError as ParsingError9 } from "@neuledge/states-parser";
var parseThisExpression = /* @__PURE__ */ __name((params, node) => {
  if (params.this?.type !== "State") {
    throw new ParsingError9(node, "ThisExpression is not allowed here");
  }
  return {
    type: "ThisExpression",
    node,
    reference: params.this
  };
}, "parseThisExpression");

// src/expression/member.ts
var parseMemberExpression = /* @__PURE__ */ __name((params, node) => {
  const properties = [
    parsePropertyExpression(params, node.property)
  ];
  while (node.object.type === "MemberExpression") {
    properties.push(parsePropertyExpression(params, node.object.property));
    node = node.object;
  }
  const object = node.object.type === "Identifier" ? parseIdentifierExpression(params, node.object) : parseThisExpression(params, node.object);
  return {
    type: "MemberExpression",
    node,
    object,
    properties
  };
}, "parseMemberExpression");
var parsePropertyExpression = /* @__PURE__ */ __name((params, node) => ({
  type: "PropertyExpression",
  node,
  name: node.name
}), "parsePropertyExpression");

// src/expression/unary.ts
var parseUnaryExpression = /* @__PURE__ */ __name((params, node) => ({
  type: "UnaryExpression",
  node,
  operator: node.operator,
  argument: parseExpression(params, node.argument)
}), "parseUnaryExpression");

// src/expression/expression.ts
var parseExpression = /* @__PURE__ */ __name((params, node) => {
  switch (node.type) {
    case "CallExpression": {
      return parseCallExpression(params, node);
    }
    case "Identifier": {
      return parseIdentifierExpression(params, node);
    }
    case "ThisExpression": {
      return parseThisExpression(params, node);
    }
    case "MemberExpression": {
      return parseMemberExpression(params, node);
    }
    case "Literal": {
      return parseLiteral(node);
    }
    case "UnaryExpression": {
      return parseUnaryExpression(params, node);
    }
    case "BinaryExpression": {
      return parseBinaryExpression(params, node);
    }
    case "LogicalExpression": {
      return parseLogicalExpression(params, node);
    }
    case "NullLiteral": {
      return parseNullLiteral(node);
    }
    default: {
      throw new TypeError(`Unexpected expression type: ${node.type}`);
    }
  }
}, "parseExpression");

// src/expression/binary.ts
var parseBinaryExpression = /* @__PURE__ */ __name((params, node) => ({
  type: "BinaryExpression",
  node,
  operator: node.operator,
  left: parseExpression(params, node.left),
  right: parseExpression(params, node.right)
}), "parseBinaryExpression");

// src/property.ts
var parseProperties = /* @__PURE__ */ __name((ctx, params, nodes) => {
  const properties = {};
  if (nodes.length) {
    for (const node of nodes) {
      const property = parseProperty(params, node);
      if (properties[property.name]) {
        throw new ParsingError10(node.key, `Duplicate property '${property.name}'`);
      }
      properties[property.name] = property;
    }
  } else {
    for (const parameter of Object.values(params.parameters)) {
      properties[parameter.name] = parseParameterProperty(params, parameter);
    }
  }
  return properties;
}, "parseProperties");
var parseProperty = /* @__PURE__ */ __name((params, node) => ({
  type: "Property",
  node,
  name: node.key.name,
  value: parseExpression(params, node.value)
}), "parseProperty");
var parseParameterProperty = /* @__PURE__ */ __name((params, parameter) => ({
  type: "Property",
  name: parameter.name,
  value: parseIdentifierExpression(params, parameter.node.key)
}), "parseParameterProperty");

// src/void.ts
var Void = {
  type: "Void",
  name: "Void"
};

// src/mutation/index.ts
var parseMutation = /* @__PURE__ */ __name((ctx, node) => {
  const target = parseTarget(ctx, node);
  const returns = parseReturns(ctx, node, target);
  const params = parseParametersContext(ctx, node.from ? target : null, node.parameters);
  const body = parseBody(ctx, params, node);
  const mutation = {
    type: "Mutation",
    node,
    ...params,
    mutation: node.from ? returns.name === Void.name ? "delete" : "update" : "create",
    name: node.key.name,
    description: node.description?.value,
    target,
    returns,
    body
  };
  applyDecorators(mutation, node.decorators, decorators3);
  return mutation;
}, "parseMutation");
var decorators3 = {
  deprecated: createDecorator(z3.object({
    reason: z3.string().optional()
  }), (state, args) => {
    state.deprecated = args.reason || true;
  })
};
var parseTarget = /* @__PURE__ */ __name((ctx, node) => {
  const targetNode = node.from ?? node.returns;
  const target = ctx.entity(targetNode.name);
  if (!target) {
    throw new ParsingError11(targetNode, `Unknown entity name '${targetNode.name}'`);
  }
  if (node.from) {
    if (target.type !== "State" && target.type !== "Either") {
      throw new ParsingError11(targetNode, `Expected state or either, got '${target.type}'`);
    }
  } else if (target.type !== "State") {
    throw new ParsingError11(targetNode, `Expected state, got '${target.type}'`);
  }
  return target;
}, "parseTarget");
var parseReturns = /* @__PURE__ */ __name((ctx, node, target) => {
  const returnsNode = node.returns;
  const returns = ctx.entity(returnsNode.name);
  if (!returns) {
    throw new ParsingError11(returnsNode, `Unknown entity name '${returnsNode.name}'`);
  }
  switch (returns.type) {
    case "State":
    case "Void": {
      break;
    }
    case "Either": {
      if (target.type !== "Either") {
        throw new ParsingError11(returnsNode, `Expected state or void, got '${returns.type}'`);
      }
      if (target.name !== returns.name) {
        throw new ParsingError11(returnsNode, `Expected either '${target.name}', got '${returns.name}'`);
      }
      break;
    }
    case "Scalar": {
      throw new ParsingError11(returnsNode, `Expected state or void, got '${returns.type}'`);
    }
  }
  return returns;
}, "parseReturns");
var parseBody = /* @__PURE__ */ __name((ctx, params, node) => parseProperties(ctx, params, node.body), "parseBody");

// src/scalar.ts
import { types } from "@neuledge/scalars";
var builtInScalars = Object.fromEntries(Object.values(types).filter((value) => value?.type === "Scalar").map((value) => [
  value.name,
  value
]));

// src/state/field.ts
import { ParsingError as ParsingError12 } from "@neuledge/states-parser";
import { z as z4 } from "zod";

// src/state/state-index.ts
var StateIndexNameRegex = /^[\w.]+$/i;
var isStateSortingIndexEquals = /* @__PURE__ */ __name((sa, ia, sb, ib) => {
  if (!ia.unique !== !ib.unique && ia.auto !== ib.auto) {
    return false;
  }
  const a_keys = Object.keys(ia.fields);
  const b_keys = Object.keys(ib.fields);
  if (a_keys.length !== b_keys.length) {
    return false;
  }
  return a_keys.every((ak, i) => {
    const bk = b_keys[i];
    return ia.fields[ak] === ib.fields[bk] && sa.fields[ak].as.entity.name === sb.fields[bk].as.entity.name && !sa.fields[ak].as.list === !sb.fields[bk].as.list && !sa.fields[ak].nullable === !sb.fields[bk].nullable;
  });
}, "isStateSortingIndexEquals");

// src/state/field.ts
var parseStateField = /* @__PURE__ */ __name((ctx, node, baseIndex) => {
  const as = parseType(ctx, node.as);
  const base = {
    node,
    name: node.key.name,
    nullable: node.nullable,
    index: node.index.value + baseIndex,
    description: node.description?.value,
    as
  };
  const field = as.entity.type === "Scalar" || !as.list ? {
    type: "ScalarField",
    ...base
  } : {
    type: "RelationField",
    ...base,
    referenceField: null
  };
  applyDecorators(field, node.decorators, decorators4);
  if (field.type === "RelationField" && !field.referenceField) {
    throw new ParsingError12(node, `Relation field '${field.name}' must have a '@reference()' field annotation`);
  }
  return field;
}, "parseStateField");
var decorators4 = {
  deprecated: createDecorator(z4.object({
    reason: z4.string().optional()
  }), (field, args) => {
    field.deprecated = args.reason || true;
  }),
  id: createDecorator(z4.object({
    sort: z4.union([
      z4.literal("asc"),
      z4.literal("desc"),
      z4.literal(1),
      z4.literal(-1)
    ]).optional(),
    auto: z4.literal("increment").optional()
  }), (field, { sort, auto }) => {
    if (field.type !== "ScalarField") {
      throw new ParsingError12(field.node, `@id can only be applied to scalar fields`);
    }
    if (field.primaryKey) {
      throw new ParsingError12(field.node, `Duplicate @id on field '${field.name}'`);
    }
    field.primaryKey = {
      name: field.name,
      fields: {
        [field.name]: sort == null || sort === "asc" || sort === 1 ? "asc" : "desc"
      },
      unique: true,
      auto
    };
  }),
  index: createDecorator(z4.object({
    sort: z4.union([
      z4.literal("asc"),
      z4.literal("desc"),
      z4.literal(1),
      z4.literal(-1)
    ]).optional(),
    unique: z4.boolean().optional(),
    name: z4.string().regex(StateIndexNameRegex).optional()
  }), (field, { sort, unique, name }) => {
    if (field.type !== "ScalarField") {
      throw new ParsingError12(field.node, `@index can only be applied to scalar fields`);
    }
    if (field.sortingIndex) {
      throw new ParsingError12(field.node, `Duplicate @index or @unique on field '${field.name}'`);
    }
    field.sortingIndex = {
      name: name || field.name,
      fields: {
        [field.name]: sort == null || sort === "asc" || sort === 1 ? "asc" : "desc"
      },
      unique
    };
  }),
  unique: createDecorator(z4.object({
    sort: z4.union([
      z4.literal("asc"),
      z4.literal("desc"),
      z4.literal(1),
      z4.literal(-1)
    ]).optional(),
    name: z4.string().regex(StateIndexNameRegex).optional()
  }), (field, { sort, name }) => {
    if (field.type !== "ScalarField") {
      throw new ParsingError12(field.node, `@unique can only be applied to scalar fields`);
    }
    if (field.sortingIndex) {
      throw new ParsingError12(field.node, `Duplicate @unique or @index on field '${field.name}'`);
    }
    field.sortingIndex = {
      name: name || field.name,
      fields: {
        [field.name]: sort == null || sort === "asc" || sort === 1 ? "asc" : "desc"
      },
      unique: true
    };
  }),
  reference: createDecorator(z4.object({
    field: z4.string().regex(/^[_a-z]\w*$/i)
  }), (field, { field: referenceField }) => {
    if (field.type === "RelationField" && field.referenceField) {
      throw new ParsingError12(field.node, `Duplicate @reference on field '${field.name}'`);
    }
    Object.assign(field, {
      type: "RelationField",
      referenceField
    });
  })
};

// src/state/state.ts
import { ParsingError as ParsingError13 } from "@neuledge/states-parser";
import { z as z5 } from "zod";
var parseState = /* @__PURE__ */ __name((node, fields, mutations, baseIndex) => {
  const state = {
    type: "State",
    node,
    name: node.id.name,
    description: node.description?.value,
    fields,
    primaryKey: {
      name: "",
      fields: {},
      unique: true
    },
    indexes: {},
    mutations,
    baseIndex
  };
  for (const field of Object.values(fields)) {
    if (field.type !== "ScalarField")
      continue;
    const { sortingIndex, primaryKey } = field;
    if (sortingIndex) {
      if (state.indexes[sortingIndex.name]) {
        throw new ParsingError13(field.node, `Index '${sortingIndex.name}' already exists`);
      }
      state.indexes[sortingIndex.name] = sortingIndex;
    }
    if (primaryKey) {
      Object.assign(state.primaryKey.fields, primaryKey.fields);
      if (primaryKey.auto) {
        state.primaryKey.auto = primaryKey.auto;
      }
    }
  }
  applyDecorators(state, node.decorators, decorators5);
  applyPrimaryKey(state, node);
  return state;
}, "parseState");
var decorators5 = {
  deprecated: createDecorator(z5.object({
    reason: z5.string().optional()
  }), (state, args) => {
    state.deprecated = args.reason || true;
  }),
  id: createDecorator(z5.object({
    fields: z5.union([
      z5.record(z5.union([
        z5.literal("asc"),
        z5.literal("desc"),
        z5.literal(1),
        z5.literal(-1)
      ])),
      z5.array(z5.string())
    ]),
    auto: z5.literal("increment").optional()
  }), (state, { fields, auto }, argsNodes) => {
    const fieldsEntries = Array.isArray(fields) ? fields.map((field) => [
      field,
      "asc"
    ]) : Object.entries(fields);
    for (const [key, sort] of fieldsEntries) {
      const field = state.fields[key];
      if (!field) {
        throw new ParsingError13(argsNodes.fields.value, `Field ${key} does not exist`);
      }
      state.primaryKey.fields[field.name] = sort === 1 || sort === "asc" ? "asc" : "desc";
    }
    if (auto) {
      state.primaryKey.auto = auto;
    }
  }),
  index: createDecorator(z5.object({
    fields: z5.union([
      z5.record(z5.union([
        z5.literal("asc"),
        z5.literal("desc"),
        z5.literal(1),
        z5.literal(-1)
      ])),
      z5.array(z5.string())
    ]),
    unique: z5.boolean().optional(),
    name: z5.string().regex(StateIndexNameRegex).optional()
  }), (state, { fields, unique, name }, argsNodes, node) => {
    const fieldsEntries = Array.isArray(fields) ? fields.map((field) => [
      field,
      "asc"
    ]) : Object.entries(fields);
    const index = {
      name: name || fieldsEntries.map(([key]) => key).join("_"),
      fields: {},
      unique
    };
    for (const [key, sort] of fieldsEntries) {
      const field = state.fields[key];
      if (!field) {
        throw new ParsingError13(argsNodes.fields.value, `Field ${key} does not exist`);
      }
      index.fields[field.name] = sort === 1 || sort === "asc" ? "asc" : "desc";
    }
    if (state.indexes[index.name]) {
      throw new ParsingError13(node, `Duplicate index name: ${index.name}`);
    }
    state.indexes[index.name] = index;
    if (index.name) {
      state.indexes[index.name] = index;
    }
  })
};
var applyPrimaryKey = /* @__PURE__ */ __name((state, node) => {
  const { primaryKey, indexes, fields } = state;
  const primaryKeys = Object.keys(primaryKey.fields);
  if (!primaryKeys.length) {
    throw new ParsingError13(node.id, "State must have at least one primary key field");
  }
  if (primaryKey.auto && primaryKeys.length > 1) {
    throw new ParsingError13(node.id, "State with auto-incrementing primary key can only have one field");
  }
  for (const key of primaryKeys) {
    const field = fields[key];
    if (field.nullable) {
      throw new ParsingError13(field.node, "Primary key field cannot be nullable");
    }
  }
  const primaryKeyName = primaryKeys.join("_");
  primaryKey.name = primaryKeyName;
  let i = 0;
  while (indexes[primaryKey.name]) {
    primaryKey.name = `${primaryKeyName}_${++i}`;
  }
  indexes[primaryKey.name] = primaryKey;
}, "applyPrimaryKey");

// src/context.ts
var ProcessingOrder;
(function(ProcessingOrder2) {
  ProcessingOrder2[ProcessingOrder2["Eithers"] = 10] = "Eithers";
  ProcessingOrder2[ProcessingOrder2["States"] = 20] = "States";
  ProcessingOrder2[ProcessingOrder2["StateRelations"] = 21] = "StateRelations";
  ProcessingOrder2[ProcessingOrder2["Mutations"] = 30] = "Mutations";
  ProcessingOrder2[ProcessingOrder2["EitherMutations"] = 31] = "EitherMutations";
  ProcessingOrder2[ProcessingOrder2["Transforms"] = 40] = "Transforms";
})(ProcessingOrder || (ProcessingOrder = {}));
var StatesContext = class {
  entityMap = {
    ...builtInScalars,
    Void
  };
  mutationMap = {};
  parent;
  processing = [];
  // iterators
  *scalars() {
    yield* this.entities("Scalar");
  }
  *states() {
    yield* this.entities("State");
  }
  *eithers() {
    yield* this.entities("Either");
  }
  *entities(type) {
    for (const key in this.entityMap) {
      const entity = this.entityMap[key];
      if (type == null || entity?.type === type) {
        yield entity;
      }
    }
  }
  // getters
  entity(name) {
    let self = this;
    do {
      const res = this.entityMap[name];
      if (res)
        return res;
    } while (self = self.parent);
    return void 0;
  }
  mutation(stateName, name) {
    const key = `${stateName || ""}.${name}`;
    let self = this;
    do {
      const res = this.mutationMap[key]?.[name];
      if (res)
        return res;
    } while (self = self.parent);
    return void 0;
  }
  // executors
  async exec(source, filepath) {
    const [document] = await this.load([
      {
        source,
        filepath
      }
    ]);
    return document;
  }
  async load(inputs) {
    const documents = inputs.map(({ source, filepath }) => parseStates(source, filepath));
    const child = new StatesContext();
    child.parent = this;
    for (const document of documents) {
      for (const item of document.body) {
        child.register(item);
      }
    }
    child.process();
    this.embed(child);
    return documents;
  }
  embed(child) {
    Object.assign(this.entityMap, child.entityMap);
    Object.assign(this.mutationMap, Object.fromEntries(Object.entries(child.mutationMap).map(([key, value]) => [
      key,
      Object.assign(this.mutationMap[key] ?? {}, value)
    ])));
  }
  // registerers
  register(node) {
    switch (node.type) {
      case "Either": {
        this.registerEither(node);
        break;
      }
      case "Scalar": {
        break;
      }
      case "State": {
        this.registerState(node);
        break;
      }
      case "Mutation": {
        this.registerMutation(node);
        break;
      }
      case "Migration": {
        break;
      }
      default: {
        throw new ParsingError14(
          node,
          // @ts-expect-error `node.type` should be never
          `Unsupported document node '${node.type}'`
        );
      }
    }
  }
  registerEntity(node, order, process) {
    const { name } = node.id;
    if (this.entity(name)) {
      throw new ParsingError14(node.id, `An entity name '${name}' already defined`);
    }
    const ref = {
      type: node.type,
      node
    };
    this.entityMap[name] = ref;
    this.processing.push({
      order,
      process: () => Object.assign(ref, process(ref))
    });
  }
  registerEither(node) {
    this.registerEntity(node, ProcessingOrder.Eithers, () => parseEither(this, node));
    this.registerEitherMutations(node);
  }
  registerState(node) {
    this.registerEntity(node, ProcessingOrder.States, (ref) => {
      const fields = {};
      const baseIndex = this.assignStateFields(node, ref, fields);
      let mutations = this.mutationMap[node.id.name];
      if (!mutations) {
        mutations = this.mutationMap[node.id.name] = {};
      }
      return parseState(node, fields, mutations, baseIndex);
    });
    this.registerStateRelations(node);
    this.registerStateTransforms(node);
  }
  registerStateRelations(node) {
    this.processing.push({
      order: ProcessingOrder.StateRelations,
      process: () => {
        const state = this.entity(node.id.name);
        for (const field of Object.values(state.fields)) {
          if (field.type !== "RelationField")
            continue;
          const { entity } = field.as;
          if (entity.type === "Scalar") {
            throw new ParsingError14(field.node, `A relation field '${field.name}' must be a state or an either`);
          }
          const states = entity.type === "State" ? [
            entity
          ] : entity.states;
          const { referenceField } = field;
          for (const state2 of states) {
            if (state2.fields[referenceField])
              continue;
            throw new ParsingError14(field.node, `A relation field '${field.name}' must reference a field '${referenceField}' in state '${state2.name}', but it does not exist`);
          }
        }
      }
    });
  }
  registerStateTransforms(node) {
    this.processing.push({
      order: ProcessingOrder.Transforms,
      process: () => {
        const state = this.entity(node.id.name);
        const mutations = this.mutationMap[node.id.name] ?? {};
        for (const mutation of Object.values(mutations)) {
          if (mutation?.mutation !== "update")
            continue;
          const returns = mutation.returns;
          if (returns.name !== state.name) {
            continue;
          }
          if (returns.type !== "State" || !isStateSortingIndexEquals(state, state.primaryKey, returns, returns.primaryKey)) {
            throw new ParsingError14(mutation.node.key, `A mutation for state '${state.name}' must return a state with the same primary key as the original state`);
          }
        }
      }
    });
  }
  registerMutation(node) {
    const form = node.from ?? node.returns;
    if (this.mutation(form.name, node.key.name)) {
      throw new ParsingError14(node.key, `The mutation name '${node.key.name}'${node.from ? ` for '${node.from.name}'` : ""} already defined`);
    }
    let entry = this.mutationMap[form.name];
    if (!entry) {
      entry = {};
      this.mutationMap[form.name] = entry;
    }
    const ref = {
      type: node.type,
      node
    };
    entry[node.key.name] = ref;
    this.processing.push({
      order: ProcessingOrder.Mutations,
      process: () => Object.assign(ref, parseMutation(this, node))
    });
  }
  registerEitherMutations(node) {
    this.processing.push({
      order: ProcessingOrder.EitherMutations,
      process: () => {
        const mutations = this.mutationMap[node.id.name] ?? {};
        for (const stateNode of node.states) {
          const state = this.entity(stateNode.name);
          let stateMutations = this.mutationMap[state.name];
          if (!stateMutations) {
            this.mutationMap[state.name] = stateMutations = {};
          }
          for (const name in mutations) {
            const mutation = mutations[name];
            if (stateMutations[name] || !mutation) {
              continue;
            }
            stateMutations[name] = {
              ...mutation,
              target: state,
              returns: !mutation.node.from || mutation.node.returns.name === mutation.node.from.name ? state : mutation.returns
            };
          }
        }
      }
    });
  }
  // processors
  process() {
    this.processing.sort((a, b) => a.order - b.order);
    for (const { process } of this.processing) {
      process();
    }
    this.processing = [];
  }
  /**
  * Assign the given `fields` dictionary with the fields of the given `node` state.
  * Return the calculated base index of the state.
  */
  assignStateFields(node, ref, fields) {
    if ("fields" in ref) {
      if (ref.baseIndex == null || ref.fields == null) {
        throw new ParsingError14(node.id, `Circular dependency detected`);
      }
      Object.assign(fields, ref.fields);
      return ref.baseIndex;
    }
    ref.fields = void 0;
    let baseIndex = 0;
    if (node.from) {
      const parent = this.entity(node.from.name);
      if (parent?.type !== "State") {
        throw new ParsingError14(node.from, `Unknown state name '${node.from.name}'`);
      }
      baseIndex += this.assignStateFields(parent.node, parent, fields) + STATE_FIELD_INDEX_MAX_INPUT_VALUE;
    }
    const sortedFields = [
      ...node.fields
    ].sort((a, b) => ("index" in a ? a.index.value : 0) - ("index" in b ? b.index.value : 0));
    for (const fieldNode of sortedFields) {
      const field = this.parseStateField(fieldNode, baseIndex);
      if (!field) {
        if (!fields[fieldNode.key.name]) {
          throw new ParsingError14(fieldNode.key, `Unknown field name '${fieldNode.key.name}'`);
        }
        delete fields[fieldNode.key.name];
        continue;
      }
      fields[field.name] = field;
    }
    ref.fields = fields;
    ref.baseIndex = baseIndex;
    return baseIndex;
  }
  parseStateField(node, baseIndex) {
    switch (node.type) {
      case "Field": {
        return parseStateField(this, node, baseIndex);
      }
      case "ReferenceField": {
        const refState = this.entity(node.state.name);
        if (refState?.type !== "State") {
          throw new ParsingError14(node.state, `Unknown state name '${node.state.name}'`);
        }
        const fields = {};
        this.assignStateFields(refState.node, refState, fields);
        const field = fields[node.key.name];
        if (!field) {
          throw new ParsingError14(node.key, `Undefined field name '${node.key.name}' on state '${node.state.name}'`);
        }
        return {
          ...field,
          index: node.index.value + baseIndex
        };
      }
      case "ExcludedField": {
        return null;
      }
    }
  }
};
__name(StatesContext, "StatesContext");
export {
  StateIndexNameRegex,
  StatesContext,
  Void,
  builtInScalars,
  isStateSortingIndexEquals,
  parseArguments,
  parseBinaryExpression,
  parseCallExpression,
  parseEither,
  parseEntity,
  parseExpression,
  parseIdentifierExpression,
  parseLiteral,
  parseLogicalExpression,
  parseMemberExpression,
  parseMutation,
  parseNonNullableEntity,
  parseNullLiteral,
  parseParameters,
  parseParametersContext,
  parseProperties,
  parseState,
  parseStateField,
  parseThisExpression,
  parseType,
  parseUnaryExpression
};
